sn_networking/src/event.rs

```diff
@@ -268,39 +270,45 @@ impl SwarmDriver {
-                        // A dead peer will cause a bunch of `OutgoingConnectionError`s
-                        // to be received within a short period.
-                        if let Some(error_count) = self.potential_dead_peers.get_mut(&peer_id) {
-                            *error_count += 1;
-                            debug!("Transport error occured, tracking potential dead_peer: {peer_id:?} - {error:?}");
-                            if *error_count > DEAD_PEER_DETECTION_THRESHOLD {
-                                trace!("Detected dead peer {peer_id:?}");
-                                self.try_trigger_replication(&peer_id, true);
-                                let _ = self.swarm.behaviour_mut().kademlia.remove_peer(&peer_id);
-                            }
-                        } else {
-                            debug!("Transport error occured, tracking potential dead_peer: {peer_id:?} - {error:?}");
-                            let _prior_none_value = self.potential_dead_peers.insert(peer_id, 1);
+                    // Related errors are: WrongPeerId, ConnectionRefused
+                    let err_string = format!("{error:?}");
+                    let is_wrong_id = err_string.contains("WrongPeerId");
+                    let is_all_connection_refused = if let DialError::Transport(ref errors) = error
+                    {
+                        errors.iter().all(|(_, error)| {
+                            let err_string = format!("{error:?}");
+                            err_string.contains("ConnectionRefused")
+                            // if let TransportError::Other(err) = error {
+                            //     // Observed on Windows is :
+                            //     // `Custom { kind: Other, error: Left(Left(Os { code: 10061, kind: ConnectionRefused, message: "..."})) }`
+                            //     trace!("err: {err:?} - {:?}", err.kind());
+                            //     let raw_os_error = err.raw_os_error();
+                            //     trace!("raw_os_error: {raw_os_error:?}");
+                            //     // Observed on Windows is
+                            //     raw_os_error == Some(10061)
+                            // } else {
+                            //     false
+                            // }
+                        })
+                    } else {
+                        true
+                    };
+
+                    if is_wrong_id || is_all_connection_refused {
+                        self.try_trigger_replication(&peer_id, true);
+                        if self
+                            .swarm
+                            .behaviour_mut()
+                            .kademlia
+                            .remove_peer(&peer_id)
+                            .is_some()
+                        {
+                            trace!("First time detected dead peer {peer_id:?}");
```

Comment:
Is this the first time? Could it not happen again?

Reply:
yes, it shall be the first time.
It will only happen again when the peer join using the same id and dropped out again.
which, autually indicates some issue of the dead peer detection algorithm, in case we don't allow such `re-join with same peer_id` happens.